### PR TITLE
Improve Prometheus query editor

### PR DESCRIPTION
**Pull request by mtanda**
Summary of changes.
* add "Legend format" field
* remove "Labels" and "Alias" field
* add Prometheus link to query editor
* expand "Query" field
* fix maxDataPoints setting

new_query_editor

### CODE REVIEW

**Comment by mtanda** [comment]
@juliusv I implement Legend format string and Prometheus link.
Would you check these changes?

**Comment by juliusv** [comment]
@mtanda Awesome, this is exciting!

I have only done a functional test so far, no code review (which might be better suited for people more familiar with Grafana, but I'm also able to help out with that if needed).

Some things I've still noticed in my testing:

- in the default legend format, if time series don't have a metric name (but
  just labels), the metric name is currently displayed as `undefined`. It
  should just be omitted in that case. You can e.g. create a metric without a
  name by querying for `rate(prometheus_local_storage_chunk_ops_total[5m])` or
  any other aggregation/transformation that drops the original metric name.
- Prometheus datasources don't work if there's a trailing slash in the
  Prometheus server URL. It'd be cool if both variants would work (slash or no
  slash).
- If a Prometheus query fails, there's no clear error displayed yet. For
  example, if I try to query "rate(foo[])", I should get the returned syntax
  error displayed somewhere (probably in the top left corner of the graph,
  where there is already a little red exclamation mark upon error, but all the
  content fields are empty).
- Maybe add a hover text to the "Link to Prometheus graph" button? Otherwise
  people might not know what the button does.

Great work!

**Comment by mtanda** [comment]
@juliusv Thanks for quick feedback!
I don't notice the case, the metric name becomes undefined.
Other feedbacks are focused on usability, its are important.
I'll fix it until today or tomorrow.

File: datasources/prometheus/datasource.js [hunk 1]
```diff
@@ -134,16 +123,25 @@ function (angular, _, kbn) {
       return { target: metricLabel, datapoints: dps };
```
Comment (jfcervera) [inline datasources/prometheus/datasource.js]: In the dps calculation function above, there is an issue,
```
return [parseInt(value[1]), value[0] * 1000];
```
The parseInt is breaking things for me, I have metrics returning decimal info and this breaks that, see example below:
```
"values":[[1429875048,"0.03"],[1429875051,"0.025188716726261422"] ....
```
I tested by removing the parseInt and all looks good after that, any reasons why you think you need the parseInt?

Thanks,
Juan

Comment (mtanda) [inline datasources/prometheus/datasource.js]: Hi Juan,

Thanks for bug report!
Sorry, I didn't consider the case, returned metric value type becomes float.

At first, I explained why I need to call parseInt here.
Grafana seems to expect plugin return datapoints as number, not string.
https://github.com/grafana/grafana/blob/380b7f4d9421369abe61b048fdf787afc4d05158/public/app/components/timeSeries.js#L90
(check the datapoint type isNumber here)
If removing the parseInt, it cause another problem (e.g. break the total value calculation)

I fix the reported bug, by using parseFloat.
grafana_parsefloat

**Comment by mtanda** [comment]
@juliusv I fix the problems except query error reporting.
It seems to be difficult to report error in query editor, I give up to implement it...
If I get a good idea, I'll try it again.

I fix the duplicate query after editing expression too.

**Comment by juliusv** [comment]
Hey, great, just testing this again.

Just noticed that when using no legend string, the special `__name__` label is displayed not only in the beginning as the metric name, but also in the "normal" labelset following the metric name. It should be excluded from there:

metric_name

**Comment by juliusv** [comment]
The other stuff (besides error reporting) looks like it's working now, great!

**Comment by mtanda** [comment]
Sorry, I break the legend string in recent change...
I fix it now.

**Comment by torkelo** [comment]
@mtanda could you rebase this, seems to be some merge conflicts

**Comment by mtanda** [comment]
@torkelo Sorry, recent changes cause conflict.

@siavashs Thanks for PR (https://github.com/grafana/grafana-plugins/pull/15), and sorry for reverting your changes in this PR.
This PR add “Legend format” mentioned at https://github.com/grafana/grafana-plugins/pull/13#issuecomment-94493739, the “Labels” field becomes obsoleted.
If you already create dashboard, you need to update panel setting, sorry…

**Comment by siavashs** [comment]
@mtanda it's ok, mine was just a quick hack :smile: 
Btw, it would be nice to have query errors in editor and/or on graphs.
Also a dropdown to select and apply prometheus functions(including tooltip about the function) would be nice.
If you can not or don't want to add these features, I will add them later.

**Comment by discordianfish** [comment]
First of all, great work @mtanda this looks really good! I'm playing with it around right now and realized that the default resolution (maxDataPoints) is set to the width of the graph in pixels.
Although this is 'correct' given that you get exactly the data to put as much information in the graph as possible, it's usually not what you want because it's really expensive and doesn't give you much more insight. That's why in the default prometheus graph endpoint this is set to 250.

What about using width/2 as default? Maybe even width/3, given the high resolution everyone has today.

**Comment by torkelo** [comment]
Why is it expensive? I have graphs that on super wide monitors,  and its super fast, and I want to use my pixels :) 

Originally graphs in grafana had maxDataPoints to pixels in graph/2  but users complained and wanted higher precision and so I increased it to so that maxDataPoints = graph width in pixels, been working great

**Comment by torkelo** [comment]
Or is it expensive fetching so many data points from Prometheus?  because Grafana metrics processing and rendering does not really care that much if there are 400 data points or 1000 data points.

**Comment by juliusv** [comment]
@torkelo Querying at a higher resolution can proportionally increase the query time in the Prometheus backend (to some extent). That might not matter at all for some queries, but will be noticeably slower for heavy ones. So the Grafana UI is still fast with multiple graphs with many lines each and 1-pixel resolution of samples?

**Comment by juliusv** [comment]
PromDash has a resolution slider for adjusting the query resolution (lowest, low, normal, high, highest). Maybe that would be a possibility here as well? But I don't know Grafana well (yet).

**Comment by juliusv** [comment]
Keep in mind also that the network traffic will become many megabytes for larger queries / dashboards, which also decreases response times over slow (e.g. wifi) networks.

**Comment by torkelo** [comment]
I have not noticed any slow down when having 10 graphs each with 5-6 series, each series about 1000 points, so grafana can handle a lot (to point, when datapoints start to go above 2000 points, its will start to hurt a little, and if you have many many series)

**Comment by juliusv** [comment]
Yeah, some users put way too many series (hundreds) in a single graph, and then have many of those graphs on one dashboard. That isn't recommended, but it'd still be great to be able to reduce the query resolution for that and the other cases.

**Comment by torkelo** [comment]
you can in grafana, in the query editors for graphite and InfluxDB you can override the default (auto handling) of the maxDataPoints value

**Comment by mtanda** [comment]
@siavashs Thanks :smile:
Graphite query editor has such dropdown.
It might be useful, I also want it!!
I'm not sure I can do it or not, I can't promise do that.
I'll consider how I can do it.

**Comment by mtanda** [comment]
@discordianfish @juliusv
I agree with your opinion, the current default value cause high load of Prometheus, the default value should be half or even less.
But, I can't change the meaning of maxDataPoints = auto, to set width of the graph in pixels, I think it should be same meaning of Graphite or other plugin.
Instead, I can add resolution option PromDash has.
By selecting low resolution, the maxDataPoints becomes half or even less, it reduce the load/response size of query.
To add resolution option, I think there are two approaches:

1. simply remove the current maxDataPoints option field from query editor, and add resolution option
2. speficy resolution in maxDataPoints field
  * If option value is number, it specify the number of points absolutely.
  * If option value is string (e.g. low, normal, high), it specify the resolution, the number of datapoints becomes half of graph pixels.
	    
If adding resolution option is acceptable, I'll implement it.

@torkelo
I want to merge this PR before additional implementation start.
Could you merge this?

**Comment by juliusv** [comment]
@mtanda I'm fine with anything that lets you lower the resolution - whatever works best with Grafana :)

By the way, I just documented this on the official Prometheus site: http://prometheus.io/docs/visualization/grafana/ (and announced it on Twitter: https://twitter.com/PrometheusIO/status/593834351269392385)

**Comment by torkelo** [comment]
@juliusv nice!

**Comment by mtanda** [comment]
@juliusv Okay, I'll think about what is best!

**Pull request merged**